src/services/providers/provider-sources/load-balance-provider.ts

```diff
@@ -0,0 +1,159 @@
+import { ChainId, Timestamp, TimeString } from '@types';
+import { IProviderSource } from '../types';
+import { chainsUnion } from '@chains';
+import { createTransport, type Transport } from 'viem';
+import ms from 'ms';
+
+export type LoadBalanceSourceConfig = LoadBalanceConfig;
+export class LoadBalanceProviderSource implements IProviderSource {
+  constructor(private readonly sources: IProviderSource[], private readonly config: LoadBalanceConfig | undefined) {
+    if (sources.length === 0) throw new Error('Need at least one source to setup the provider source');
+  }
+
+  supportedChains() {
+    return chainsUnion(this.sources.map((source) => source.supportedChains()));
+  }
+
+  getViemTransport({ chainId }: { chainId: ChainId }) {
+    const transports = this.sources
+      .filter((source) => source.supportedChains().includes(chainId))
+      .map((source) => source.getViemTransport({ chainId }));
+    if (transports.length === 0) throw new Error(`Chain with id ${chainId} not supported`);
+    return loadBalance(transports, this.config);
+  }
+}
+
+function loadBalance(transports_: readonly Transport[], config: LoadBalanceConfig = {}): Transport {
+  const { minSuccessRate = 0.05, minSamples = 3, maxAttempts = 3, samplesTtl = '30m' } = config;
+
+  const rpcMetrics: MethodMetrics[] = transports_.map(() => ({}));
+
+  return ({ chain, timeout, ...rest }) => {
+    const transports = transports_.map((t) => t({ chain, timeout, ...rest }));
+    return createTransport({
+      key: 'load-balance',
+      name: 'Load Balancing',
+      async request({ method, params }) {
+        let availableTransports = transports.map((transport, index) => ({ transport, index }));
+        if (maxAttempts) availableTransports.splice(maxAttempts);
+        const processRequest = async (): Promise<any> => {
```

Comment:
Is it necessary to declare the function and call it immediately? Why not execute the code directly?